tracer/src/Datadog.Trace/Telemetry/Collectors/ConfigurationTelemetryCollector.cs

```diff
@@ -0,0 +1,175 @@
+            if (reconfigureCount != 1)
+            {
+                SetHasChanges();
+                return;
+            }
+
+            _azureApServicesMetadata = appServicesMetadata;
+            var host = HostMetadata.Instance;
+            _hostData = new HostTelemetryData
+            {
+                ContainerId = ContainerMetadata.GetContainerId(),
+                Os = FrameworkDescription.Instance.OSPlatform,
+                OsVersion = Environment.OSVersion.ToString(),
+                Hostname = host.Hostname,
+                KernelName = host.KernelName,
+                KernelRelease = host.KernelRelease,
+                KernelVersion = host.KernelVersion,
+            };
+
+            _isTracerInitialized = true;
+            SetHasChanges();
+        }
+
+        public void RecordSecuritySettings(SecuritySettings securitySettings)
+        {
+            _securitySettings = securitySettings;
+            SetHasChanges();
+        }
+
+        public bool HasChanges()
+        {
+            return _isTracerInitialized && _hasChangesFlag == 1;
+        }
+
+        /// <summary>
+        /// Get the application data. Will be null if not yet initialized.
+        /// </summary>
+        public ApplicationTelemetryData GetApplicationData()
+        {
+            var settings = _settings;
+            return settings?.ApplicationData;
+        }
+
+        /// <summary>
+        /// Get the application data. Will be null if not yet initialized.
+        /// </summary>
+        public HostTelemetryData GetHostData()
+        {
+            return _hostData;
+        }
+
+        /// <summary>
+        /// Get the latest data to send to the intake.
+        /// </summary>
+        /// <returns>Null if there are no changes, or the collector is not yet initialized</returns>
+        public ConfigTelemetryData GetConfigurationData()
+        {
+            var hasChanges = Interlocked.CompareExchange(ref _hasChangesFlag, 0, 1) == 1;
+            if (!_isTracerInitialized || !hasChanges)
+            {
+                return null;
+            }
+
+            var settings = _settings.Settings;
+
+            var data = new ConfigTelemetryData
+            {
+                Platform = FrameworkDescription.Instance.ProcessArchitecture,
+                Enabled = settings.TraceEnabled,
+                AgentUrl = settings.Exporter.AgentUri.ToString(),
+                Debug = GlobalSettings.Source.DebugEnabled,
+#pragma warning disable CS0618
+                AnalyticsEnabled = settings.AnalyticsEnabled,
+#pragma warning restore CS0618
+                SampleRate = settings.GlobalSamplingRate,
+                SamplingRules = settings.CustomSamplingRules,
+                LogInjectionEnabled = settings.LogsInjectionEnabled,
+                RuntimeMetricsEnabled = settings.RuntimeMetricsEnabled,
+                RoutetemplateResourcenamesEnabled = settings.RouteTemplateResourceNamesEnabled,
```

Comment:
I suggest matching the casing:

```suggestion
                RoutetemplateResourcenamesEnabled = settings.RouteTemplateResourceNamesEnabled,
```

Reply:
I just realized that the casing you used is so that the resulting JSON field would be `routetemplate_resourcenames_enabled` which is much easier to read than `route_template_resource_names_enabled`. The current approach seems good to me

Reply:
Yeah, that was definitely the reason :stare:

Reply:
The telemetry spec has changed now actually, so we need to explicitly use the serialized names here anyway, which reduces the layer of abstraction here, and makes it clearer what the resulting data will look like. Done in https://github.com/DataDog/dd-trace-dotnet/pull/2241/commits/6abf33e5c72eae47f234cedf10c432d8f8160f0b